lib/Parser/Parse.cpp

```diff
@@ -11445,11 +11445,23 @@ void Parser::FinishDeferredFunction(ParseNodeBlock * pnodeScopeList)
     uint saveNextBlockId = m_nextBlockId;
     m_nextBlockId = pnodeScopeList->blockId + 1;
 
+    ParseContext parseContext;
+    this->CaptureContext(&parseContext);
+
     FinishFunctionsInScope(pnodeScopeList,
-        [this](ParseNodeFnc * pnodeFnc)
+        [this, &parseContext](ParseNodeFnc * pnodeFnc)
     {
         Assert(pnodeFnc->nop == knopFncDecl);
 
+        // We need to scan this function based on the already known limits of the function declaration as some of
+        // the state such as fAllowIn may not be available at this point. Some of this state depends on the context
+        // of the function declaration. For example, a function declaration may be inside a for..in statement's var
+        // declaration. It may not be appropriate/possible to try and save all such context information. Functions
+        // that actually get deferred achieve this by going through the ParseSourceWithOffset code path.
+        this->GetScanner()->Clear();
+        this->GetScanner()->SetText(parseContext.pszSrc, pnodeFnc->cbMin /*+ this->m_scan.m_cMinTokMultiUnits*/, pnodeFnc->LengthInBytes(), pnodeFnc->ichMin, parseContext.isUtf8, parseContext.grfscr, pnodeFnc->lineNumber);
+        this->GetScanner()->Scan();
```

Comment:
I'm curious: since the scanner state is reset here, is there any danger that callers of FinishDeferredFunction function will assume that the scanner state has not been reset?

Reply:
Good call. I see there is some minimal code in finish parse block code after this that may depend on the scanner state. I have restored the parser state at the end of FinishDeferredFunction to set the scanner state to what it was before.